### PR TITLE
[2.8][WebProfilerBundle] Fix bundle usage in Content-Security-Policy context

### PR DESCRIPTION
**Pull request by romainneutron**
| Q | A |
| --- | --- |
| Branch? | 2.8 |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #15397 |
| License | MIT |
| Doc PR | N/A |

Hello, this PR fixes the compatibility of the WebprofilerBundle in a context where Content-Security-Policy headers are could prevent `unsafe-inline` of `script-src` or `style-src` directives.

I intentionally open this PR against 2.8 instead of 2.3. Fixing this in 2.3/2.7 is a PITA due to the legacy codebase of the WebProfiler. It would imply doing the job twice. Moreover, doing this is not trivial at all. I hope you all understand.

Job is done at 80%, so I open the PR to get feedback.
- [x] Add tests to the `ContentSecurityPolicyHandler` and `WebDebugToolbarListener`

### CODE REVIEW

File: src/Symfony/Bundle/WebProfilerBundle/CSP/ContentSecurityPolicyHandler.php [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+            'csp_script_nonce' => $this->generateNonce(),
+            'csp_style_nonce' => $this->generateNonce(),
+        ], $nonces);
+
+        $scriptRule = false;
+        $styleRule = false;
+
+        $headers = $this->getCSPHeaders($response);
+
+        foreach($headers as $header => $directives) {
```
Comment (hhamon) [inline src/Symfony/Bundle/WebProfilerBundle/CSP/ContentSecurityPolicyHandler.php]: missing space after `foreach`.

File: src/Symfony/Bundle/WebProfilerBundle/CSP/ContentSecurityPolicyHandler.php [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+
+    /**
+     * Updates Content-Security-Policy headers in a response.
+     *
+     * @param Response $response
+     * @param array    $nonces
+     *
+     * @return array
+     */
+    private function updateCSPheaders(Response $response, array $nonces = array()) {
```
Comment (stof) [inline src/Symfony/Bundle/WebProfilerBundle/CSP/ContentSecurityPolicyHandler.php]: CS issue

File: src/Symfony/Bundle/WebProfilerBundle/CSP/ContentSecurityPolicyHandler.php [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+     * Retrieves the Content-Security-Policy headers (either X-Content-Security-Policy or Content-Security-Policy) from
+     * a response.
+     *
+     * @param Response $response The response
+     *
+     * @return array An associative array of headers
+     */
+    private function getCSPHeaders(Response $response)
+    {
+        $headers = [];
```
Comment (stof) [inline src/Symfony/Bundle/WebProfilerBundle/CSP/ContentSecurityPolicyHandler.php]: using the short array syntax is forbidden in Symfony

File: src/Symfony/Bundle/WebProfilerBundle/CSP/ContentSecurityPolicyHandler.php [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+     *
+     * @param Request  $request
+     * @param Response $response
+     *
+     * @return array
+     */
+    public function getNonces(Request $request, Response $response)
+    {
+        if ($request->headers->has('X-WebProfiler-Script-Nonce') && $request->headers->has('X-WebProfiler-Style-Nonce')) {
+            return [
```
Comment (stloyd) [inline src/Symfony/Bundle/WebProfilerBundle/CSP/ContentSecurityPolicyHandler.php]: This should be `array()` as 2.8 still supports PHP 5.3

File: src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php [hunk 5]
```diff
@@ -40,8 +41,9 @@ class WebDebugToolbarListener implements EventSubscriberInterface
     protected $mode;
     protected $position;
     protected $excludedAjaxPaths;
+    protected $cspHandler;
```
Comment (stof) [inline src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php]: should be private (any new property is private)

File: src/Symfony/Bundle/WebProfilerBundle/Resources/config/profiler.xml [hunk 6]
```diff
@@ -8,6 +8,7 @@
         <parameter key="web_profiler.controller.profiler.class">Symfony\Bundle\WebProfilerBundle\Controller\ProfilerController</parameter>
         <parameter key="web_profiler.controller.router.class">Symfony\Bundle\WebProfilerBundle\Controller\RouterController</parameter>
         <parameter key="web_profiler.controller.exception.class">Symfony\Bundle\WebProfilerBundle\Controller\ExceptionController</parameter>
+        <parameter key="web_profiler.csp_handler.class">Symfony\Bundle\WebProfilerBundle\CSP\ContentSecurityPolicyHandler</parameter>
```
Comment (stof) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/config/profiler.xml]: we don't introduce any new class parameter anymore

File: src/Symfony/Bundle/WebProfilerBundle/CSP/ContentSecurityPolicyHandler.php [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+                'csp_script_nonce' => $request->headers->get('X-WebProfiler-Script-Nonce'),
+                'csp_style_nonce' => $request->headers->get('X-WebProfiler-Style-Nonce'),
+            ];
+        }
+
+        if ($response->headers->has('X-WebProfiler-Script-Nonce') && $response->headers->has('X-WebProfiler-Style-Nonce')) {
```
Comment (stloyd) [inline src/Symfony/Bundle/WebProfilerBundle/CSP/ContentSecurityPolicyHandler.php]: Those lines are duplicates of `if` above.

File: src/Symfony/Bundle/WebProfilerBundle/Resources/config/profiler.xml [hunk 8]
```diff
@@ -32,6 +34,9 @@
             <argument>%kernel.debug%</argument>
         </service>
 
+        <service id="web_profiler.csp_handler" class="%web_profiler.csp_handler.class%">
```
Comment (stof) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/config/profiler.xml]: should be a private service

File: src/Symfony/Bundle/WebProfilerBundle/CSP/ContentSecurityPolicyHandler.php [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+     *  - They are otherwise randomly generated
+     *
+     * @param Request  $request
+     * @param Response $response
+     *
+     * @return array
+     */
+    public function getNonces(Request $request, Response $response)
+    {
+        if ($request->headers->has('X-WebProfiler-Script-Nonce') && $request->headers->has('X-WebProfiler-Style-Nonce')) {
```
Comment (javiereguiluz) [inline src/Symfony/Bundle/WebProfilerBundle/CSP/ContentSecurityPolicyHandler.php]: What if we rename `X-WebProfiler-*` headers as `X-SymfonyProfiler-*` ?

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Icon/symfony.svg [hunk 10]
```diff
@@ -1,4 +1,4 @@
-<svg version="1.1" xmlns="http://www.w3.org/2000/svg" x="0px" y="0px" width="24" height="24" viewBox="0 0 24 24" enable-background="new 0 0 24 24" xml:space="preserve">
+<svg version="1.1" xmlns="http://www.w3.org/2000/svg" x="0px" y="0px" width="40" height="40" viewBox="0 0 24 24" enable-background="new 0 0 24 24" xml:space="preserve">
```
Comment (stof) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Icon/symfony.svg]: this change is broken. Please don't revert other PRs

File: src/Symfony/Bundle/WebProfilerBundle/CSP/ContentSecurityPolicyHandler.php (hunk 4 above)
Comment (javiereguiluz) [inline src/Symfony/Bundle/WebProfilerBundle/CSP/ContentSecurityPolicyHandler.php]: The use of `[ ]` instead of `array()` is forbidden in Symfony (and 2.8 requires PHP 5.3.* anyway).

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig [hunk 11]
```diff
@@ -41,6 +41,8 @@
                 options.maxTries = options.maxTries || 0;
                 xhr.open(options.method || 'GET', url, true);
                 xhr.setRequestHeader('X-Requested-With', 'XMLHttpRequest');
+                xhr.setRequestHeader('X-WebProfiler-Script-Nonce', options.scriptNonce);
```
Comment (stof) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig]: should be added only when the options are provided

File: src/Symfony/Bundle/WebProfilerBundle/CSP/ContentSecurityPolicyHandler.php [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+                $scriptRule = true;
+                $headers[$header]['script-src'][] = sprintf('\'nonce-%s\'', $nonces['csp_script_nonce']);
+            }
+            if ($this->CSPpreventsUnsafeInline($directives, 'style-src')) {
+                $styleRule = true;
+                $headers[$header]['style-src'][] = sprintf('\'nonce-%s\'', $nonces['csp_style_nonce']);
+            }
+        }
+
+        if (!$styleRule && !$scriptRule) {
```
Comment (javiereguiluz) [inline src/Symfony/Bundle/WebProfilerBundle/CSP/ContentSecurityPolicyHandler.php]: Looking at the code of this method, I think you could replace these two flags (`$styleRule` and `$scriptRule`) by a single flag.

**Comment by stof** [comment]
IMO, this should be considered as a new feature rather than a bugfix

File: src/Symfony/Bundle/WebProfilerBundle/CSP/ContentSecurityPolicyHandler.php [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+    }
+
+    /**
+     * Converts a directives set array into Content-Security-Policy header.
+     *
+     * @param array $directives The directives set
+     *
+     * @return string The Content-Security-Policy header
+     */
+    private function generateCSPHeader(array $directives)
```
Comment (javiereguiluz) [inline src/Symfony/Bundle/WebProfilerBundle/CSP/ContentSecurityPolicyHandler.php]: Given that this method just reduces an array to a string, maybe we could use the `array_reduce()` method.

**Comment by stof** [comment]
thus, another reason to consider it a new feature is that any bundle extending the profiler will have to consider it too.

File: src/Symfony/Bundle/WebProfilerBundle/Controller/ProfilerController.php [hunk 14]
```diff
@@ -397,4 +400,18 @@ protected function getTemplateManager()
 
         return $this->templateManager;
     }
+
+    private function renderWithCSPNonces(Request $request, $template, $variables, $code = 200, $headers = array('Content-Type' => 'text/html'))
+    {
+        $response = new Response('', $code, $headers);
+
+        $nonces = $this->cspHandler ? $this->cspHandler->getNonces($request, $response) : [];
```
Comment (javiereguiluz) [inline src/Symfony/Bundle/WebProfilerBundle/Controller/ProfilerController.php]: Forbidden short array notation in this line

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Icon/symfony.svg (hunk 10 above)
Comment (javiereguiluz) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Icon/symfony.svg]: This change is unrelated ... and there is a pending PR about this.

File: src/Symfony/Bundle/WebProfilerBundle/Tests/EventListener/WebDebugToolbarListenerTest.php [hunk 15]
```diff
@@ -30,7 +32,7 @@ public function testInjectToolbar($content, $expected)
 
         $response = new Response($content);
 
-        $m->invoke($listener, $response, Request::create('/'));
+        $m->invoke($listener, $response, Request::create('/'), ['csp_script_nonce' => 'scripto', 'csp_style_nonce' => 'stylo']);
```
Comment (javiereguiluz) [inline src/Symfony/Bundle/WebProfilerBundle/Tests/EventListener/WebDebugToolbarListenerTest.php]: Short array syntax here.

**Comment by javiereguiluz** [comment]
I have a global concern about this code. As usually happens with acronyms, it's hard to name methods and variables. But right now there are different criteria:

```php
updateCSPheaders()
getCSPHeaders()
$cspHandler;
```

I propose to treat `CSP` as `Csp` word, so:

```php
updateCspHeaders()
getCspHeaders()
$cspHandler;
```

File: src/Symfony/Bundle/WebProfilerBundle/Resources/config/profiler.xml [hunk 16]
```diff
@@ -32,6 +34,9 @@
...
+        </service>
```
Comment (hhamon) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/config/profiler.xml]: Can be removed and replaced with a single `<service/>` monotag.

Comment (javiereguiluz) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/config/profiler.xml]: @hhamon I think they are called "empty tags" ... but I absolutely love the "monotag" word :grin:

Comment (hhamon) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/config/profiler.xml]: I never knew how to call them and this is why I also end up calling them "monotag" :D

**Comment by romainneutron** [comment]
@stof I proposed this as a bug fix because it prevents usage of Symfony in a content-security-policy context. This is fully BC with other usage (Silex and any bundle extending it)
Bundles extending this one should only care about this in case they are not CSP compatible (some of them are already compatible) and want to be

**Comment by romainneutron** [comment]
For the record, I updated the PR according to your comments

File: src/Symfony/Bundle/WebProfilerBundle/Controller/ProfilerController.php [hunk 17]
```diff
@@ -397,4 +400,18 @@ protected function getTemplateManager()
 
         return $this->templateManager;
     }
+
+    private function renderWithCspNonces(Request $request, $template, $variables, $code = 200, $headers = array('Content-Type' => 'text/html'))
```
Comment (stof) [inline src/Symfony/Bundle/WebProfilerBundle/Controller/ProfilerController.php]: according to your talk at the SymfonyLive right now, wouldn't it be better to use signatures instead ?

Comment (romainneutron) [inline src/Symfony/Bundle/WebProfilerBundle/Controller/ProfilerController.php]: Actually this PR provides CSP support for setups where `unsafe-inline` is disabled in **development** environment (because it only affects the WebProfilerBundle that should not be enabled in prod in any case)

So, even if it's recommended to use signature instead of nonce, it's a recommendation for production or staging. In development, it enables user to fully uses test with their CSP and uses the WebProfilerBundle.

So, no, it's not a problem :)

**Comment by romainneutron** [comment]
I just amended the PR according to Javier's comment about array_reduce. It's now ready, feedback are welcome :)

**Comment by romainneutron** [comment]
A side note argument about "new feature" vs "bug fix":

Considering this a "new feature" would disable the WebProfilerBundle to any user that uses Content Security Policy in their Symfony Application until 3.2. Moreover, it would mean that Symfony would not provide a LTS version compatible with Content Security Policy before the end of 2017 (3.4 if I'm not wrong).
I definitely think it's a bug fix because Content Security Policies are a standard supported by most of the modern browsers (http://caniuse.com/#feat=contentsecuritypolicy2) and used by many applications, GitHub included.
Last but not least, I consider this one of the key feature to prevent XSS, and it should not be unusable with Symfony.

**Comment by romainneutron** [comment]
Replaced by #18568 that proposes this has a new feature in 3.2
